chat-plugins/uno.js

```diff
@@ -0,0 +1,694 @@
+	}
+
+	join(user) {
+		if (this.started) return false;
+		let ip = user.latestIp;
+		let userid = user.userid;
+
+		if (this.playerCap && this.list.length >= this.playerCap) {
+			return user.sendTo(this.room, "The game is already full.");
+		}
+
+		if (userid in this.data || ip in this.joinedIps) {
+			return user.sendTo(this.room, "You have already have an alt joined.");
+		}
+
+		this.list.push(userid);
+		//init empty hand
+		this.data[userid] = [];
+		this.joinedIps[ip] = 1;
+		//make join messages clear away after game starts
+		this.room.add("|uhtml|init" + this.room.unoGameId + "|" + user.name + " has joined the game.");
+	}
+
+	start(user) {
+		if (this.started) return false;
+		if (this.list.length < 2) return user.sendTo(this.room, "There aren't enough players!");
+		this.started = true;
+		this.room.add("The game has started!");
+		//hide the start message
+		this.room.add("|uhtmlchange|new" + this.room.unoGameId + "|<div class=\"infobox\">The game has started.</div>");
+		//get first player
+		this.player = this.list[~~(Math.random() * this.list.length)];
+		this.room.add("The first player is: " + getUserName(this.player) + ".");
+
+		//create the deck
+		this.deck = shuffleDeck(initDeck(this.list.length));
+
+		//deal the cards
+		this.list.forEach(function (u) {
+			this.giveCard(u, 7);
+		}.bind(this));
+
+		//get the top card and put it into discard pile
+		while (!this.top || this.top.charAt(0) === "Z") {
+			this.top = this.deck.shift();
+			this.discard.push(this.top);
+		}
+		//declare the top card
+		this.room.add("|uhtml|" + this.room.unoGameId + "card" + this.id + "|" + "The top card is " + buildCard(this.top, true));
+		this.uhtmlChange = "|uhtmlchange|" + this.room.unoGameId + "card" + this.id + "|The top card is <b>" + getCardName(this.top, true) + "</b>.";
+
+		this.applyEffects();
+
+		this.initNextTurn();
+	}
+
+	giveCard(userid, number, display) {
+		if (!number) number = 1;
+		let newCards = [];
+		for (let i = 0; i < number; i++) {
+			let newCard = this.deck.shift();
+			this.data[userid].push(newCard);
+			newCards.push(newCard);
+			//shuffle deck if deck is empty
+			if (this.deck.length === 0) {
+				if (this.discard.length === 0) {
+					//if no cards in discard pile, add a new deck.
+					this.discard = initDeck(1);
+				}
+				//shuffle discard pile to create new deck for drawing
+				this.deck = shuffleDeck(this.discard);
+			}
+		}
+		if (display && Users(userid)) {
+			return Users(userid).sendTo(this.room, "|raw|You have drawn the following card(s): " + buildHand(newCards, true));
+		}
+		//only for draw 1
+		if (newCards.length === 1) return newCards[0];
+	}
+
+	getNextPlayer(number) {
+		//make list run over if it's at the last player
+		let playerList = this.list.concat(this.list);
+		//get index of current player
+		let index = playerList.indexOf(this.player);
+		//apply next player
+		this.player = playerList[index + (number || 1)];
+	}
+
+	applyEffects() {
```

Comment:
If you're applying the card effects like this, then why even contruct the entire (hardcoded) deck in 2 char strings?